File: Lintcode 041 - Maximum Subarray.java
Lintcode 41 - Maximum Subarray.java

public class Solution {
    //方法1：Greedy
    public int maxSubArray(int[] nums) {
        // write your code
        if(nums == null || nums.length == 0) {
            return 0;
        }
        
        int sum = 0;
        int max = Integer.MIN_VALUE;
        for(int i = 0; i < nums.length; i++) {
            sum += nums[i];
            max = Math.max(max, sum);
            sum = Math.max(sum, 0);
        }
        
        return max;
    }
}

File: Lintcode 041 - Maximum Subarray.cs
using System;
using System.Collections;

namespace CSharp
{
    class Program{
        static void Main()
        {
            int[] A = new int[]{-2, 2, -3, 4, -1, 2, 1, -5, 3};

            int result = MaxiSub(A);
            Console.WriteLine(result);

        }

        static int MaxiSub(int[] A){
            int max = int.MinValue;
            int sum = 0;

            for(int i = 0; i < A.Length; i++)
            {
                sum += A[i];
                max = Math.Max(max, sum);
                sum = Math.Max(0, sum);
            }

            return max;
        }
    }

}
